### PR TITLE
Enforce more Scalastyle rules and add MiMa binary compatibility checks

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch adds MiMa binary compatibility checks and enables a number of Scalastyle rules (after fixing the existing violations).

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -73,3 +73,25 @@ ScoverageSbtPlugin.ScoverageKeys.coverageHighlighting := {
...
+
+mimaDefaultSettings ++ Seq(
+  previousArtifact := Some("com.databricks" %% "spark-csv" % "1.2.0"),
+  binaryIssueFilters ++= Seq(
+    // These classes are not intended to be public interfaces:
+    ProblemFilters.excludePackage("com.databricks.spark.csv.CsvRelation"),
+    ProblemFilters.excludePackage("com.databricks.spark.csv.util.InferSchema"),
+    ProblemFilters.excludePackage("com.databricks.spark.sql.readers"),
+    // We allowed the private `CsvRelation` type to leak into the public method signature:
+    ProblemFilters.exclude[IncompatibleResultTypeProblem](
```
Comment (JoshRosen) [inline build.sbt]: This is an example of why it's important to enforce explicit return types in public methods: the inferred type was CsvRelation which, as far as I know, was not intended to be a public class.

**Comment by falaki** [comment]
Why moving ```CsvReader``` to ```readers/readers```?

**Comment by JoshRosen** [comment]
@falaki, its package did not correspond to its filename / location.

**Comment by falaki** [comment]
Code changes LGTM. Please take a look at the test compilation failure in ```ReaderSuite.scala```

**Comment by JoshRosen** [comment]
Updated, so this is now passing tests.

**Pull request closed without merging**